### ISSUE TITLE
Insufficient permission to manipulate /usr/local/bin

### ISSUE THREAD

**Issue by twillowman** [rancher-sandbox/rancher-desktop#1155]
### Rancher Desktop Version

0.7.1

### Rancher Desktop K8s Version

1.21.7 (stable)

### What operating system are you using?

macOS

### Operating System / Build Version

macOS Monterey 12.1

### What CPU architecture are you using?

arm64 (Apple Silicon)

### Actual Behavior

In Rancher Desktop, Supporting Utilities I have the error `Insufficient permission to manipulate /usr/local/bin: Error: EACCES: permission denied, access '/usr/local/bin'` and the box next to each supporting utility is greyed out. 

<img width="1052" alt="image" src="https://user-images.githubusercontent.com/5553208/147143139-d436851c-a901-4824-ab27-bc2aaadd9b52.png">

This affects only installing supporting utilities from Rancher Desktop. All other functions of Rancher Desktop seem to be working as expected. The utilities work and can interact with the running Kubernetes instance when installed any other way. 

### Steps to Reproduce

This is a new laptop, first time configuration. 

I downloaded Rancher.Desktop-0.7.1.aarch64.dmg from the releases page, opened it, dragged "Rancher Desktop.app" into Applications, and launched it. 

It went through the first-time Kubernetes download without issue. Quitting and re-opening Rancher Desktop has not made any difference. 

### Result

<img width="1052" alt="Screen Shot 2021-12-22 at 12 09 08 PM" src="https://user-images.githubusercontent.com/5553208/147144010-633e3158-99a7-4646-a56e-2ba89853c4d9.png">

Logs attached in case they're helpful.
[background.log](https://github.com/rancher-sandbox/rancher-desktop/files/7764743/background.log)
[images.log](https://github.com/rancher-sandbox/rancher-desktop/files/7764744/images.log)
[k3s.log](https://github.com/rancher-sandbox/rancher-desktop/files/7764745/k3s.log)
[k8s.log](https://github.com/rancher-sandbox/rancher-desktop/files/7764746/k8s.log)
[lima.ha.stderr.log](https://github.com/rancher-sandbox/rancher-desktop/files/7764747/lima.ha.stderr.log)
[lima.ha.stdout.log](https://github.com/rancher-sandbox/rancher-desktop/files/7764748/lima.ha.stdout.log)
[lima.log](https://github.com/rancher-sandbox/rancher-desktop/files/7764749/lima.log)
[lima.serial.log](https://github.com/rancher-sandbox/rancher-desktop/files/7764750/lima.serial.log)
 

### Expected Behavior

I expect the Supporting Utilities tab of Rancher Desktop to be able to install the supporting utilities it lists.

**Comment by jasjisdo** [comment]
IHMO in this case rancher should also ask for root permissions to install the _Supporting Utilities_.
After the initial installation, it also asked for root privileges to perform the k3s network configuration.
When installing the Supporting Utilities, rancher Desktop should behave analogously and not inconsistently.

**Comment by mattfarina** [comment]
@twillowman thanks for reporting the issue and suggesting a solution.

**Comment by spuder** [comment]
I'm also running into this on 0.7.0 on OSX Big Sur. 

Where _should_ the symlinks point to so we can manually work around it? I'd really like to have `docker` symlinked to my path. 

https://github.com/rancher-sandbox/rancher-desktop/blob/7c0ad62de6db286982f30c2902935f85c29a48cc/src/k8s-engine/lima.ts#L661-L673

https://github.com/rancher-sandbox/rancher-desktop/blob/f8e59c96b1f297c132826b93f33a163368bb4209/src/k8s-engine/unixlikeIntegrations.ts#L100-L124

**Comment by jandubois** [comment]
@spuder The problem is the `/usr/local/bin` no longer exists, or is not owned by yourself. You can fix that by running the following commands in a terminal:

```shell
sudo mkdir -p /usr/local/bin
sudo chown $USER /usr/local/bin
```

The first `sudo` command will prompt you for your password. Once the directory exists, and is writable by you, RD should be able to create the symlinks.

**Comment by spuder** [comment]
Thank you. I found that `/usr/local/bin` _does_ exist on my new M1 mac running Big Sur, but it was owned by root

Running this worked around the issue for me: 

```
sudo chown $USER /usr/local/bin
```

![Screen Shot 2021-12-30 at 10 34 24 PM](https://user-images.githubusercontent.com/242382/147805405-a320e87b-1360-4ed4-b5fc-61b09ae1f7b6.png)

**Comment by enceem** [comment]
had the same issue but 
`sudo chown $USER /usr/local/bin`
resolved it

**Comment by mgudmund** [comment]
IMHO /usr/local/bin should be owned by root. So changing it to be owned by $USER is a workaround, not a fix. I would think the solution proposed by @jasjisdo is the right way to do it.

**Comment by fungiboletus** [comment]
I agree that `usr/local/bin` should be owned by root. My workaround is to run
```bash
sudo chown $USER /usr/local/bin
```
then I can check the checkboxes in rancher desktop,
and I restore the root owner:
```
sudo chown root /usr/local/bin
```

**Comment by weaversam8** [comment]
Just to add a data point, this issue is still present in 1.0.0-beta1.

**Comment by abs1010** [comment]
sudo chown $USER /usr/local/bin worked for me. Thanks

**Comment by istvandesign** [comment]
I confirm that it reproduces on Mac OS Monterey Version 12.1 on a MacBook Pro with M1 Pro I have the same issue after installing rancher desktop.

**Comment by boogi** [comment]
The same in version 1.0.0-beta1 on Monterey 12.1. Temporarily changing owner workaround worked for me also.

**Comment by selfisekai** [comment]
also to note: if installed from Homebrew, the symlinks are [generated by Homebrew](https://github.com/Homebrew/homebrew-cask/blob/e4c1513a02fbaa6814aab2c4b3d8e6a612ec9da3/Casks/rancher.rb#L29-L33), but the app doesn't reflect that state.

```
==> Downloading https://github.com/rancher-sandbox/rancher-desktop/releases/download/v0.7.1/Rancher.Desktop-0.7.1.aarch64.dmg
Already downloaded: /Users/lauren.liberda/Library/Caches/Homebrew/downloads/962f40e4104948f7fc3fc8ba3fb2f4548a9a0257b5a86fc006d4ee0af31f75b2--Rancher.Desktop-0.7.1.aarch64.dmg
==> Installing Cask rancher
==> Moving App 'Rancher Desktop.app' to '/Applications/Rancher Desktop.app'
==> Linking Binary 'helm' to '/opt/homebrew/bin/helm'
==> Linking Binary 'kim' to '/opt/homebrew/bin/kim'
==> Linking Binary 'nerdctl' to '/opt/homebrew/bin/nerdctl'
==> Linking Binary 'kubectl' to '/opt/homebrew/bin/kubectl'
==> Linking Binary 'docker' to '/opt/homebrew/bin/docker'
🍺  rancher was successfully installed!
```

`/opt/homebrew/bin/docker` is a symlink to `'/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/docker'`, and so on

**Comment by ssheraliev** [comment]
It seems that "mucked up" permissions issue on /usr/local/bin directory.
I am on M1 chip and with latest v.1.0.0 version:
<img width="941" alt="Screenshot 2022-01-27 at 16 46 10" src="https://user-images.githubusercontent.com/15828558/151393140-94e832a7-189d-42ef-b0a1-4d743f5ad4a6.png">
Fix would be; adding your user-account into admin group:
sudo chown $(whoami):admin /usr/local/bin

**Comment by jt-nti** [comment]
I worked around the issue by creating the symlinks manually...

```
sudo ln -s "/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/docker" /usr/local/bin/docker
sudo ln -s "/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/helm" /usr/local/bin/helm
sudo ln -s "/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/kubectl" /usr/local/bin/kubectl
sudo ln -s "/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/nerdctl" /usr/local/bin/nerdctl
```

**Comment by ex-git** [comment]
For Homebrew user, this is the work around:

```
sudo ln -s "/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/docker" /opt/homebrew/bin/docker
sudo ln -s "/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/helm" /opt/homebrew/bin/helm
sudo ln -s "/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/kubectl" /opt/homebrew/bin/kubectl
sudo ln -s "/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/nerdctl" /opt/homebrew/bin/nerdctl
```

**Comment by hcguersoy** [comment]
> also to note: if installed from Homebrew, the symlinks are [generated by Homebrew](https://github.com/Homebrew/homebrew-cask/blob/e4c1513a02fbaa6814aab2c4b3d8e6a612ec9da3/Casks/rancher.rb#L29-L33), but the app doesn't reflect that state.
> 
> ```
> ==> Downloading https://github.com/rancher-sandbox/rancher-desktop/releases/download/v0.7.1/Rancher.Desktop-0.7.1.aarch64.dmg
> Already downloaded: /Users/lauren.liberda/Library/Caches/Homebrew/downloads/962f40e4104948f7fc3fc8ba3fb2f4548a9a0257b5a86fc006d4ee0af31f75b2--Rancher.Desktop-0.7.1.aarch64.dmg
> ==> Installing Cask rancher
> ==> Moving App 'Rancher Desktop.app' to '/Applications/Rancher Desktop.app'
> ==> Linking Binary 'helm' to '/opt/homebrew/bin/helm'
> ==> Linking Binary 'kim' to '/opt/homebrew/bin/kim'
> ==> Linking Binary 'nerdctl' to '/opt/homebrew/bin/nerdctl'
> ==> Linking Binary 'kubectl' to '/opt/homebrew/bin/kubectl'
> ==> Linking Binary 'docker' to '/opt/homebrew/bin/docker'
> 🍺  rancher was successfully installed!
> ```
> 
> `/opt/homebrew/bin/docker` is a symlink to `'/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/docker'`, and so on

This seems to be changed. Installing RD 1.0.1 with brew doesn't create these links:

```
> brew install rancher 
==> Downloading https://github.com/rancher-sandbox/rancher-desktop/releases/download/v1.0.1/Rancher.Desktop-1.0.1.aarch6
==> Downloading from https://objects.githubusercontent.com/github-production-release-asset-2e65be/306701996/d88f1bff-32e
######################################################################## 100.0%
==> Installing Cask rancher
==> Moving App 'Rancher Desktop.app' to '/Applications/Rancher Desktop.app'
🍺  rancher was successfully installed!
```

Checking `/opt/homebrew/bin` shows that these links doesn't exist.
I'm on Apple Silicon, too.

**Comment by ab-mohamed** [comment]
I did the following workaround steps:

1. After installing Rancher Desktop, I ran the following script:
```
#!/bin/bash
for BINARY in docker helm kubectl nerdctl
do
  sudo ln -s /Applications/Rancher\ Desktop.app/Contents/Resources/resources/darwin/bin/${BINARY} /usr/local/bin
  sudo chown -h $USER /usr/local/bin/${BINARY}
done
```

2.  The new symbolic links look like the following:
```
$ ls -lh /usr/local/bin
total 0
lrwxr-xr-x  1 ab    wheel    80B Feb  6 17:24 docker -> /Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/docker
lrwxr-xr-x  1 ab    wheel    78B Feb  6 17:24 helm -> /Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/helm
lrwxr-xr-x  1 ab    wheel    81B Feb  6 17:24 kubectl -> /Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/kubectl
lrwxr-xr-x  1 ab    wheel    81B Feb  6 17:24 nerdctl -> /Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/nerdctl
```
And all supporting utilities are checked even if you do not have proper permission for `/usr/local/bin` directory:
<img width="937" alt="image" src="https://user-images.githubusercontent.com/43576598/152690810-1619407d-b704-4b38-937b-6456788e7589.png">

3. `kubectl` command, for example, works fine now:
```
$ kubectl get nodes
NAME                   STATUS   ROLES                  AGE     VERSION
lima-rancher-desktop   Ready    control-plane,master   2d18h   v1.22.6+k3s1
```

Note that you have to remove the symbolic links in the `/usr/local/bin` directory manually if you remove Rancher Desktop.

**Comment by ppihus** [comment]
Lots of bad workaround recommended here. Please do not `chown` your `/usr/local/bin` to your `$USER`, but if you indeed decide this is the only way, then please change it back as soon as possible.

**Comment by hcguersoy** [comment]
In such cases I prefer to add the path in the `.zshrc` or, in this case, defining aliases in the same file pointing to the specific binaries.

**Comment by ssheraliev** [comment]
> Lots of bad workaround recommended here. Please do not `chown` your `/usr/local/bin` to your `$USER`, but if you indeed decide this is the only way, then please change it back as soon as possible.

Could you explain why and how would you fix instead of workaround?

**Comment by jandubois** [comment]
Note that all the Rancher Desktop shortcuts go to the same directory, so you could put that directory in your `PATH` in the `~/.profile` corresponding to your shell. This doesn't even require `sudo`.

This is what happens with the Windows version[^1]; the directory is added at the end of the `PATH`, so the user can override the bundled version by installing them into a directory earlier on the `PATH`.

You can also put the name of that directory into a `/etc/pathd.d` file, and for the supported shells, it will be put on the `PATH` automatically. Creating the file requires `sudo` though:

```console
$ cat /etc/paths.d/rancher-desktop
/Applications/Rancher\ Desktop.app/Contents/Resources/resources/darwin/bin
```

Note that you will have to deal with shadowing on your `PATH` yourself then. E.g. if you also install `Docker Desktop`, it may put symlinks into `/usr/local/bin` any time you start it (via its privileged helper process), which may then hide the versions from Rancher Desktop.

It's kind of hard to find a one-size-fits-all solution for this issue.

[^1]: On Windows we actually put both the Windows and the Linux versions of the utilities at the end of the `PATH`. Since the Windows versions have an extra `.exe` file extension, they don't shadow each other, and both Windows command prompts, and WSL shells will automatically invoke the right version.

**Comment by ppihus** [comment]
> > Lots of bad workaround recommended here. Please do not `chown` your `/usr/local/bin` to your `$USER`, but if you indeed decide this is the only way, then please change it back as soon as possible.
> 
> Could you explain why...

That's really Linux/Unix administration 101 and goes against basic security practices. The permissions of `/usr/local/bin` are set that way for a reason. Going into more detail would be offtopic in this thread, but you can read more from the links below.

- https://askubuntu.com/a/955528
- https://unix.stackexchange.com/a/37726
- https://apple.stackexchange.com/questions/1393/are-my-permissions-for-usr-local-correct 

> ... and how would you fix instead of workaround?

I don't have a proper fix at the moment, but installing those tools manually, even with Homebrew, seems to be a more sane workaround. Unfortunately I just upgraded my MacOS to 12.3 Beta, which seems to have completely broken Rancher, so I'm unable to give a step-by-step instructions for this workaround.

**Comment by mateka** [comment]
As a workaround, I have run Rancher Desktop from a terminal as a root. Then, I have installed the missing tools, stopped Rancher Desktop and run it normally.

**Comment by mateka** [comment]
I have started Rancher Desktop as a normal user and created lima VM. They I have started it as a root (starting VM failed), installed nerdctl and switched back to normal user. I could not used brew, because there is no nerdctl for Apple Silicon. 

I am using brew for other tools, but /usr/local/bin is still owned by root. In FAQ (https://docs.brew.sh/FAQ) they say, that brew uses MacOS sandbox and do not chown /usr/local/bin to $USER.

**Comment by ppihus** [comment]
Brew does not change the ownership of `/usr/local/bin` to `$USER`. Maybe it did at some point, but not currently.

**Comment by jandubois** [comment]
> Brew does not change the ownership of `/usr/local/bin` to `$USER`. Maybe it did at some point, but not currently.

Did you actually try it? I just did (on a pristine VM running Big Sur), and you'll be surprised to see this:

```console
jan@Jans-Mac ~ % ls -la /usr/local
total 0
drwxr-xr-x   2 root  wheel   64  1 Jan  2020 .
drwxr-xr-x@ 11 root  wheel  352  1 Jan  2020 ..
jan@Jans-Mac ~ % /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
==> Checking for `sudo` access (which may request your password)...
Password:

[...]

==> Next steps:
- Run brew help to get started
- Further documentation:
    https://docs.brew.sh

jan@Jans-Mac ~ % ls -la /usr/local                                                                              
total 0
drwxr-xr-x  14 root  wheel  448  8 Feb 10:48 .
drwxr-xr-x@ 11 root  wheel  352  1 Jan  2020 ..
drwxrwxr-x   2 jan   admin   64  8 Feb 10:48 Caskroom
drwxrwxr-x   2 jan   admin   64  8 Feb 10:48 Cellar
drwxrwxr-x   2 jan   admin   64  8 Feb 10:48 Frameworks
drwxr-xr-x  22 jan   admin  704  8 Feb 10:51 Homebrew
drwxrwxr-x   3 jan   admin   96  8 Feb 10:51 bin
drwxrwxr-x   3 jan   admin   96  8 Feb 10:54 etc
drwxrwxr-x   2 jan   admin   64  8 Feb 10:48 include
drwxrwxr-x   2 jan   admin   64  8 Feb 10:48 lib
drwxrwxr-x   2 jan   admin   64  8 Feb 10:48 opt
drwxrwxr-x   2 jan   admin   64  8 Feb 10:48 sbin
drwxrwxr-x   6 jan   admin  192  8 Feb 10:54 share
drwxrwxr-x   3 jan   admin   96  8 Feb 10:48 var
```

When you think about it, how could `brew install ...` install something into `/usr/local/bin` unless it is writable by the user? Brew is not using `sudo` to install anything.

Note that I said Intel Macs; the install location is different on M1 Macs, and on Linux, so obviously it won't change the ownership of `/usr/local/bin` if it doesn't write to it.

**Comment by ppihus** [comment]
I stand corrected. Did not realise there is such difference between Apple Silicon vs Intel.

**Comment by victor-v-rad** [comment]
My workaround for RD [1.0.1](https://github.com/rancher-sandbox/rancher-desktop/releases/tag/v1.0.1): 
1. [Factory reset ](https://docs.rancherdesktop.io/getting-started/features/#factory-reset)
2. Delete files described here https://github.com/rancher-sandbox/rancher-desktop/issues/1128
3. `sudo chown $USER /usr/local/bin`
4. Reinstall Rancher Desktop. Now it asks for permissions again and creates necessary links in /usr/local/bin
5. `sudo chown root /usr/local/bin`

**Comment by igwejk** [comment]
@mattfarina would you accept a PR for this I gave it a shot?

**Comment by lazyfrosch** [comment]
I would suggest this permission change, or ask for sudo to manipulate these links.

    sudo chgrp staff /usr/local/bin
    sudo chmod g+w /usr/local/bin

**Comment by pboushy** [comment]
> I would suggest this permission change, or ask for sudo to manipulate these links.
> 
> ```
> sudo chgrp staff /usr/local/bin
> sudo chmod g+w /usr/local/bin
> ```

I highly recommend against that. That would grant ALL users on the Mac access to write to /usr/local/bin.
That’d allow any attacker/malware to write to a spot that users would run with fairly little protection.

Setting the group to admin provides similar protection as wheel without requiring sudo.

**Comment by lazyfrosch** [comment]
> Setting the group to admin provides similar protection as wheel without requiring sudo.

I was under the assumption that `staff` is an admin group

**Comment by thumperward** [comment]
Changing the permissions on a MacOS system directory is categorically not the right approach here. Could people please stop spamming this issue with inane workarounds?

**Comment by jandubois** [comment]
@thumperward Please temper your language a little. While some workarounds are less desirable than others, there is no need to insult people who are trying to help.

Please see https://github.com/rancher-sandbox/rancher-desktop/issues/1155#issuecomment-1032135332 for workarounds that don't involve running Rancher Desktop as `root` (you absolutely should not be doing that), or changing the ownership of `/usr/local/bin`, which could have security implications (that some users find acceptable).

**Comment by jandubois** [comment]
> I was under the assumption that `staff` is an admin group

`staff` is the group of **all** locally defined users (there many non-user local accounts that are not in `staff`). It is a subset of `everybody`, which also includes accounts defined by directory services.

The `admin` group is assigned to all accounts that have `admin` rights.

**Comment by jarrellmark** [comment]
This could be fixed by linking to the binaries in `/opt/rancher-desktop/bin` and adding `/opt/rancher-desktop/bin` to the PATH. `/opt/rancher-desktop/bin` exists after installing Rancher Desktop.

1. Install Rancher Desktop
2. `% cd /opt/rancher-desktop/bin`
3. `% sudo ln -s "/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin"/* .`
4. `printf "\n# Rancher Desktop\nexport PATH=\"/opt/rancher-desktop/bin:\$PATH\"\n\n" >> ~/.zshrc`

This way `/usr/local/bin` doesn't need to be touched.

I'd recommend Rancher Desktop do something like this by default. It's similar to how mambaforge and homebrew work.

For example:

```
% ls -l /opt
                                                                                 
total 0
drwxr-xr-x  32 jarrellmark  admin  1024 Feb 22 19:23 homebrew
drwxr-xr-x   4 root      wheel   128 Mar 21 12:27 rancher-desktop
```

```
% cat ~/.zshrc

# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/mjarrell/mambaforge/bin/conda' 'shell.zsh' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/mjarrell/mambaforge/etc/profile.d/conda.sh" ]; then
        . "/Users/mjarrell/mambaforge/etc/profile.d/conda.sh"
    else
        export PATH="/Users/mjarrell/mambaforge/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<
```

**Comment by Mazorius** [comment]
They could also user ~/bin to symlink the executables instead of provide them system wide. But do not know the impact then :) So maybe this is not a good solution. I like the approach of @jarrellmark THANKS!

**Comment by jandubois** [comment]
> This could be fixed by linking to the binaries in `/opt/rancher-desktop/bin` and adding `/opt/rancher-desktop/bin` to the PATH. `/opt/rancher-desktop/bin` exists after installing Rancher Desktop.

This is pretty close to what we will be doing in the next release: All binaries will be sym-linked into `~/.rd/bin`, and the app will offer to add that directory to your `PATH` via your shell profile (should support at least `bash`, `fish`, `zsh`).

You can of course decline and do this yourself, if you want more control over where to place the directory relative to your other directories on the `PATH`.

**Comment by ouksal** [comment]
I wish ranger-docker could work!

**Comment by opensource21** [comment]
> Simple one-liner for those who were looking for it based on the suggestion here: [#1155 (comment)](https://github.com/rancher-sandbox/rancher-desktop/issues/1155#issuecomment-1032135332)
> 
> ```
> echo '/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin' | sudo tee /etc/paths.d/rancher-desktop
> ```

I think additional a `sudo chmod o+r /etc/paths.d/rancher-desktop` is necessary.

**Comment by adamkpickering** [comment]
This has been addressed. The fix will be included in the upcoming 1.3.0. For more information please see #1732.
